src/Sylius/Bundle/UserBundle/Context/CustomerContext.php

```diff
@@ -23,27 +25,58 @@ class CustomerContext implements CustomerContextInterface
     /**
      * @var SecurityContextInterface
      */
-    private $securityContext;
+    protected $securityContext;
+
+    /**
+     * @var SessionInterface
+     */
+    protected $session;
 
     /**
      * @param SecurityContextInterface $securityContext
+     * @param SessionInterface $session
      */
-    public function __construct(SecurityContextInterface $securityContext)
-    {
+    public function __construct(
+        SecurityContextInterface $securityContext,
+        SessionInterface $session
+    ) {
         $this->securityContext = $securityContext;
+        $this->session = $session;
     }
 
     /**
-     * Gets customer based on currently logged user.
+     * Gets customer based on currently logged user or from session (if user is not authenticated).
      *
      * @return CustomerInterface|null
      */
     public function getCustomer()
     {
-        if ($this->securityContext->getToken() && $this->securityContext->isGranted('IS_AUTHENTICATED_REMEMBERED')
-            && $this->securityContext->getToken()->getUser()
-        ) {
-            return $this->securityContext->getToken()->getUser()->getCustomer();
+        $user = $this->getUser();
+
+        if (null !== $user) {
+            return $user->getCustomer();
+        }
+
+        return $this->session->get('customer');
+    }
+
+    /**
+     * Stores customer in session.
+     *
+     * @param CustomerInterface $customer
+     */
+    public function setCustomer(CustomerInterface $customer)
+    {
+        $this->session->set('customer', $customer);
```

Comment:
What if someone logs out and then logs in as different user on same browser*? :)

\* I know it's not really possible but still curious :wink: